php-classes/Slate/CBL/StudentCompetency.php

```diff
@@ -53,16 +63,334 @@ class StudentCompetency extends \ActiveRecord
         ]
     ];
 
-    public static function isCurrentLevelComplete($Student, $Competency)
+    public static $dynamicFields = [
+        'demonstrationsLogged' => [
+            'getter' => 'getDemonstrationsLogged'
+        ],
+        'demonstrationsComplete' => [
+            'getter' => 'getDemonstrationsComplete'
+        ],
+        'demonstrationsAverage' => [
+            'getter' => 'getDemonstrationsAverage'
+        ],
+        'demonstrationData' => [
+            'getter' => 'getDemonstrationData'
+        ],
+        'effectiveDemonstrationsData' => [
+            'getter' => 'getEffectiveDemonstrationsData'
+        ]
+
+    ];
+
+    public function calculateStartingRating($autoSave = true)
     {
-        $completion = $Competency->getCompletionForStudent($Student);
+        $validSkills = DB::allRecords(
+            'SELECT ID, '.
+            ' CAST( '.
+                'REPLACE( '.
+            '       IFNULL ( '.
+            '           JSON_EXTRACT(DemonstrationsRequired, CONCAT(\'$."\', %u, \'"\')), '.
+            '           JSON_EXTRACT(DemonstrationsRequired, "$.default")'.
+            '       ), '.
+            '   \'"\', ' .
+            '   "")'.
+            '   AS UNSIGNED '.
+            ' ) AS DemonstrationsRequirements '.
+            '  FROM `%s` '.
+            ' WHERE CompetencyID = %u '.
+            'HAVING DemonstrationsRequirements > 0 ',
+            [
+                $this->Level,
+                Skill::$tableName,
+                $this->Competency->ID
+            ]
+        );
+        $validSkillIds = array_column($validSkills, 'ID');
+
+        $ratedSkills = DB::oneValue(
+
+            'SELECT SUM(IF(skillRatings.total > 0, 1, 0)) as totalRatings '.
+            '  FROM ( '.
+                    'SELECT COUNT(*) AS total '.
+            		'  FROM `%1$s` ds '.
+            		'  JOIN `%2$s` d '.
+            		'	ON d.ID = ds.DemonstrationID '.
+            		' WHERE ds.SkillID IN (%3$s) '.
+            		'   AND ds.TargetLevel = %4$u '.
+            		'   AND ds.DemonstratedLevel > 0 '.
+            		'   AND d.StudentID = %5$u '.
+            		'   AND ds.Override = false '.
+            		' GROUP BY SkillID '.
+            ' ) AS skillRatings ',
+            [
+                DemonstrationSkill::$tableName,
+                Demonstration::$tableName,
+                join(', ', $validSkillIds),
+                $this->Level,
+                $this->StudentID
+            ]
+        );
+
+        // check if each skill has been rated atleast once for the current level
+        if ((int)$ratedSkills !== count($validSkills)) {
```

Comment:
This check isn't sufficient, there could be ratings for skills with 0 demonstrations required. Use `DB::valuesTable` to get an array of totals indexed by skill ID, and then foreach through `$validSkillIds` and return null when you encounter one where `empty($skillRatings[$skillId])`